### DISCUSSION TITLE
Auto-converting single to double quotes?

### DISCUSSION THREAD

**Discussion by hmbakhsh** [zed-industries/zed#7951] [Feature Requests]
When I save a file my single quotes are automatically converted to double quotes, is there any way to disable this? Can't seem to find anything which fixes this.

https://github.com/zed-industries/zed/assets/105169800/e432f3a8-5e4b-44dc-a382-087e5b98f9ce

**Comment by d1y** [comment]
disable save auto format
```
"format_on_save": "off"
```

**Reply by hmbakhsh** [reply]
Thank you, appreciate the help. Any way to specifically target the single → double quote formatting? Quite like the auto semi-colon insertion.

**Reply by d1y** [reply]
The zed js/ts default format tool use prettier, so you can add the configuration file yourself, refer to: https://prettier.io/docs/en/options#quotes

And. The config-filename rule:

![image.png](https://s2.loli.net/2024/02/17/suYP8jzQo1U2JNn.png)

**Reply by SomeoneToIgnore** [reply]
Zed itself only supports toggling specific formatters on and off.
Here's the default settings.json abstract:

```
// How to perform a buffer format. This setting can take 4 values:
  //
  // 1. Format code using the current language server:
  //     "formatter": "language_server"
  // 2. Format code using an external command:
  //     "formatter": {
  //       "external": {
  //         "command": "prettier",
  //         "arguments": ["--stdin-filepath", "{buffer_path}"]
  //       }
  //     }
  // 3. Format code using Zed's Prettier integration:
  //     "formatter": "prettier"
  // 4. Default. Format files using Zed's Prettier integration (if applicable),
  //    or falling back to formatting via language server:
  //     "formatter": "auto"
  "formatter": "auto",
```

Assuming that your settings are default, this would mean that you have `  "formatter": "auto",` enabled, which would try to use prettier formatting on languages that are applicable, then fall back to the language server formatting.

`*.js` files are applicable for prettier formatting, and Zed will first try to find a prettier installed in the project (by looking into package.json and node_modules) — if found, it will use project configs for that prettier to format things (set in prettier-specific files or the same package.json, in prettier-specific section).
If fails to find any installed prettiers, Zed installs a "default" prettier that would be used to format such files — in that case, `"prettier"` section can be used to pass any prettier options (should be exactly the same as one would configure prettier via package.json)

Interestingly, https://prettier.io/docs/en/options.html says that single quotes are not the default setting, so something in your project spoils it?

And, lastly, if the prettier formatting is not enabled, Zed uses the corresponding LSP server to format documents — then that server has to be configured somehow to avoid such formatting, but those config are server-specific and hard to tell without more context.

**Reply by SomeoneToIgnore** [reply]
See also https://github.com/zed-industries/zed/discussions/7917 that seem to suggest the LSP option for such formatting, if I understood it correctly.

**Comment by Amparose** [comment]
For me, I was able to turn it off in my Zed settings.json:

```json
{
  "languages": {
    "JavaScript": {
      "prettier": {
        "singleQuote": true
      }
    }
  }
}
```

There are blocks in the Default settings JSON that can be referred to for precise targeting.

**Reply by dushmanta05** [reply]
Could you please help me if there is any work around to auto save to single quote even if there's double quote it'll be single when saved? Thank you.

**Reply by Ptujec** [reply]
I found that I needed to add `"allowed": true` otherwise prettier was not used anymore. 

```
"JavaScript": {
      "prettier": {
        "singleQuote": true,
        "allowed": true
      }
    }
```

**Reply by chibuzoro-salesfloor** [reply]
this is what worked for me while using php. I had to include the language_servers, and entire prettier default options. I had hoped I did not have to repeat the php language config in its entirety and only needed to add the singleQuote option. 
```
"languages": {
    "PHP": {
      "language_servers": ["phpactor", "!intelephense", "..."],
      "prettier": {
        "allowed": true,
        "plugins": ["@prettier/plugin-php"],
        "parser": "php",
        "singleQuote": true
      }
    }
  }
  ```

**Reply by xplosionmind** [reply]
Any ideas on how to make single quotes the default in HTML?

**Reply by dandv** [reply]
> ```json
> {
>   "languages": {
>     "JavaScript": {
>       "prettier": {
>         "singleQuote": true
> ```

I don't see any `single...` autocomplete suggestions under the `prettier` block for TypeScript:

<img width="438" height="284" alt="image" src="https://github.com/user-attachments/assets/5451be1c-b995-44a2-bb7c-46549ffdd1ee" />

**Comment by dandee** [comment]
Can Zed detect the desired quote setting from eslintrc or the current context, and if it cannot then f**k off with not doing anything in this area please? Is it difficult?

**Comment by markoffden** [comment]
It's December 2025 and I still wonder what does Zed use to format such aggressively under the hood? I have my eslint setup (standard from Vite+React project), prettier disallowed in settings, prettier allowed in settings with `"singleQuote": true` etc. But this dude keeps massacring my code and the only way is to turn off formatting on save.

```
    "languages": {
        "Vue.js": {
            "code_actions_on_format": {
                "source.fixAll.eslint": true
            },
            "prettier": {
                "allowed": false
            },
            "formatter": []
        },
        "JavaScript": {
            "code_actions_on_format": {
                "source.fixAll.eslint": true
            },
            "prettier": {
                "singleQuote": true,
                "allowed": true
            },
            "formatter": []
        }
    }
```

**Reply by markoffden** [reply]
Ok, I figured out the issue was me :)

You have to specifically mention 'TSX', 'TypeScript` etc. as language which apparently I didn't do. Also you may put those per-language instructions in settings root to apply them for all languages as a general rule.

**Reply by juliasser** [reply]
Hey, did you figure out how to achieve this for Typescript?

I added the above to my settings.json, replacing "JavaScript" with "TypeScript" but I can't get it to work...thought I'll ask since you mentioned typescript :)

```
{
  "languages": {
    "TypeScript": {
      "code_actions_on_format": {
        "source.fixAll.eslint": true,
      },
      "prettier": {
        "singleQuote": true,
        "allowed": true,
      },
      "formatter": [],
    },
  },
}
```

**Reply by juliasser** [reply]
Ugh, sorry, I am stupid, I was working in a tsx file, not a typescript file. Using this to auto-convert back to single quotes worked.

```
{
  "languages": {
    "TSX": {
      "prettier": {
        "singleQuote": true,
        "allowed": true,
      },
    },
  },
}
```

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
